content/blog/e2e-design/index.md

```diff
@@ -0,0 +1,276 @@
+- `--command`: a command to run after the environment is started, this may be useful when users need to install some extra tools or apply resources from command line, like `istioctl install --profile=demo`;
+- `--wait`: wait until the given condition is met; the most frequently-used strategy should be `for=service-health`, that makes the `e2e setup` command to wait for all services to be `healthy`; other possible strategies may be something like `for="log:Started Successfully"`, `for="http:localhost:8080/healthcheck"`, etc. if really needed;
+
+
+### Trigger Inputs
+
+```shell
+swctl e2e trigger --interval=3s --times=0 --action=http --url="localhost:8080/users"
+swctl e2e trigger --interval=3s --times=0 --action=cmd --cmd="curl localhost:8080/users"
+swctl e2e trigger # If configuration file e2e.yaml is present
+```
+
+- `--interval=3s`: trigger the action every 3 seconds;
+- `--times=0`: how many times to trigger the action, `0=infinite`;
+- `--action=http`: the action of the trigger, i.e. "perform an http request as an input";
+- `--action=cmd`: the action of the trigger, i.e. "execute the `cmd` as an input";
+
+
+### Query Output
+
+```shell
+swctl service ls
+```
+
+this does exactly the same as what `swctl` is doing at present;
+
+
+### Verify
+
+```shell
+swctl e2e verify --actual=actual.data.yaml --expected=expected.data.yaml
+swctl e2e verify --query="service ls" --expected=expected.data.yaml
+swctl e2e verify # If configuration file e2e.yaml is present
+```
+
+- `--actual`: the actual data file;
+- `--expected`: the expected data file;
+- `--query`: the query to get the actual data, will be executed in command `swctl`;
+  > the `--query` option will get the output into a temporary file and use the `--actual` under the hood;
+
+
+### Cleanup
+
+This step is automatically done by the framework according to the setup step.
+
+### Summarize
+
+To summarize, the directory structure of a test case might be
+
+```text
+case-name
+├── agent-service        # optional, an arbitrary project that is used in the docker-compose.yaml if needed
+│   ├── Dockerfile
+│   ├── pom.xml
+│   └── src
+├── docker-compose.yaml
+├── e2e.yaml             # see a sample below
+└── testdata
+    ├── expected.endpoints.service1.yaml
+    ├── expected.endpoints.service2.yaml
+    └── expected.services.yaml
+```
+
+or
+
+```text
+case-name
+├── kind.yaml
+├── bookinfo
+│   ├── bookinfo.yaml
+│   └── bookinfo-gateway.yaml
+├── e2e.yaml             # see a sample below
+└── testdata
+    ├── expected.endpoints.service1.yaml
+    ├── expected.endpoints.service2.yaml
+    └── expected.services.yaml
+```
+
+a sample of `e2e.yaml` may be
+
+```yaml
+setup:
+  env: kind
+  file: kind.yaml
+  resources: bookinfo.yaml
+  command: | # it can be a shell script or anything executable
+    istioctl install --profile=demo -y
+    kubectl label namespace default istio-injection=enabled
+
+  # OR
+  # env: compose
+  # file: docker-compose.yaml
+
+  wait: 
+    for: service-health
```

Comment:
I have questions about the "wait" segment:
Are we able to merge those wait step into one independent block , if there're multiple prerequisites to start the test?
Plus, the "for" syntax seems to be a little bit ambiguous since many of users initially recognize this as some sort of iteration thing

Reply:
> Are we able to merge those wait step into one independent block , if there're multiple prerequisites to start the test?

https://github.com/apache/skywalking-website/pull/172#discussion_r545571540 this suggestion should address your concern, if we have corresponding `wait` for each resource, one should configure a `wait` for each resource and after all of the resources are `wait`ed, things should be ready, do you have scenarios where one resource needs 1+ prerequisites?

> Plus, the "for" syntax seems to be a little bit ambiguous since many of users initially recognize this as some sort of iteration thing

suggestion? `until`? "wait for" is a grammar I think,

Reply:
[#172 (comment)](https://github.com/apache/skywalking-website/pull/172#discussion_r545571540) is a great integrated solution and it fulfills the step-by-step logic, and that solve my question. thanks for pointing out to me.  
As for the naming, until is definitely a much clearer option in my view